<a href="https://colab.research.google.com/github/Marlon-Sbardelatti/machine-learning/blob/feature%2Ftrabalho-04/trabalho-04/neural_network.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

> **Equipe**: 01<br>
> **Estudantes**: Cristina Siewert Jansen, Marlon Sbardelatti e Sofia Sousa Lindner

# **REDES NEURAIS ARTIFICIAIS**

# PARTE 1: Classificação Binária - Diagnóstico de Câncer de Mama

Base de dados: Breast Cancer Wisconsin (Diagnostic) -
https://archive.ics.uci.edu/dataset/17/breast+cancer+wisconsin+diagnostic

In [2]:
# Instala bilbiotecas Keras e Tensorflow
!pip install scikeras==0.13.0

In [3]:
!pip install tensorflow==2.21.0 scikit-learn==1.9.0

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 572.6/572.6 MB 3.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.1/9.1 MB 106.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 108.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 327.1/327.1 kB 26.7 MB/s eta 0:00:00
  Attempting uninstall: protobuf
    Found existing installation: protobuf 5.29.6
    Uninstalling protobuf-5.29.6:
      Successfully uninstalled protobuf-5.29.6
  Attempting uninstall: h5py
    Found existing installation: h5py 3.16.0
    Uninstalling h5py-3.16.0:
      Successfully uninstalled h5py-3.16.0
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.6.1
    Uninstalling scikit-learn-1.6.1:
      Successfully uninstalled scikit-learn-1.6.1
  Attempting uninstall: tensorflow
    Found existing installation: tensorflow 2.20.0
    Uninstalling tensorflow-2.20.0:
      Successfully uninstalled tensorflow-2.20.0
ERROR: pip's dependency re

In [1]:
# após instalar as bibliotecas, é necessário reiniciar o notebook
# Para isso, rode esta célula sozinha antes de prosseguir.
exit()

In [1]:
import pandas as pd
import tensorflow as tf
import sklearn
import scikeras

In [2]:
# Exibe versões das bibliotecas instaladas
pd.__version__, tf.__version__, sklearn.__version__, scikeras.__version__

('2.2.2', '2.21.0', '1.9.0', '0.13.0')

### Carregar os dados

**1) Carregue a base de dados, faça a divisição de treino e teste (para isso, utilize a função train_test_split do sklearn), como o tamanho da base de teste de 0.25.**

In [3]:
from sklearn.model_selection import train_test_split

# Lê arquivos CSV
breast_data = pd.read_csv('dados_breast.csv')
breast_labels = pd.read_csv('rotulos_breast.csv')

# Converte para ndarray em X e y
X = breast_data.values
y = breast_labels.values

# Separa dataset em conjunto de treino e teste (25%)
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.25,
)

# Exibe dimensões do conjunto de treino
X_train.shape, y_train.shape

((426, 30), (426, 1))

In [4]:
# Exibe dimensões do conjunto de teste
X_test.shape, y_test.shape

((143, 30), (143, 1))

Ambos os conjuntos têm duas dimensões: número de amostras, número de colunas. No caso do eixo X, as 30 colunas representam features distintas das amostras de entrada. No caso do eixo y, há apenas uma coluna, que é associada ao rótulo de cada amostra.

A separação de 426 amostras para treino e 143 amostras para teste confirma a separação escolhida de 25%, visto que 426 + 143 = 569 (total) e 569 × 0.25 ≈ 143.

### Estrutura da Rede Neural Artificial e Teste

In [5]:
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import cross_val_score
from tensorflow.keras.models import Sequential
import tensorflow as tf
from tensorflow.keras import backend as k

**2) Crie a RNA com as seguintes configurações:**

**a) uma camada de entrada de com 30 neurônios;**

**b) uma camada oculta densa com 16 neurônios. Porque a camada oculta tem essa quantidade de neurônios?**

**c) adicione na camada oculta a função de ativação relu e inicialize os pesos de utilize o Random uniform initializer (https://www.tensorflow.org/api_docs/python/tf/keras/initializers);**

**d) adicione a camada de saíde com a função de ativação sigmoid. Porque foi utilizada esta função para a saída da RNA?**

In [6]:
from tensorflow.keras.layers import Input, Dense
from tensorflow.keras.initializers import RandomUniform

ENTRY_LAYER_SIZE = 30
HIDDEN_LAYER_SIZE = 16

# d) Inicializador dos pesos aleatórios da camada oculta
initializer = RandomUniform(minval=0.0, maxval=1.0)

neural_network = Sequential([
    # a) Camada de Entrada
    Input(shape=(ENTRY_LAYER_SIZE,)),

    # b) Camada Oculta
    Dense(units=HIDDEN_LAYER_SIZE, activation='relu', kernel_initializer=initializer),

    # d) Camada de Saída: Sigmoid para classificação binária
    Dense(units=1, activation='sigmoid')
])

classifier = KerasClassifier(
    model=neural_network,
    verbose=0
)

**b) uma camada oculta densa com 16 neurônios. Porque a camada oculta tem essa quantidade de neurônios?**<br>
Porque o número de neurônios da camada oculta é: nº de neurônios de entrada + nº de saídas / 2.<br>
Dessa forma, temos: (30 + 2) / 2 = 32 / 2 = 16.


**d) adicione a camada de saída com a função de ativação sigmoid. Porque foi utilizada esta função para a saída da RNA?**<br>
Porque a classificação é binária, ou seja, só existem duas classes possíveis: 0 (Sem câncer) ou 1 (Com câncer). Sendo assim, a função sigmoide por padrão utiliza o limiar 0.5, o que permite a classificação em exatas duas classes.

**3) Porque utilizamos a classe chamada Sequential para a RNA?**<br>
A classe Sequential permite a definição de camadas sequenciais à rede, o que altera completamente sua arquitetura e comportamento, visto que o resultado de cada camada é utilizado na próxima. Ou seja, para definirmos que os resultados perpassam camada de entrada, oculta e saída sequencialmente, precisamos utilizar o Sequential.

**4) A partir da RNA gerada, explique o que são os valores apresentados na tabela da rede_neural.summary()**<br>
São apresentadas as camadas oculta e de saída da rede, bem como a quantidade de parâmetros e saídas de cada uma. Naturalmente, a camada oculta produz 16 saídas porque contém 16 neurônios, enquanto a camada de saída produz apenas uma: a classe.<br>
Já a coluna de parâmetros exibe quantos parâmetros treináveis existem na camada. Dado que o cálculo dos parâmetros é (nº de entradas x nº de neurônios) + biases, a camada oculta tem 496 pois (31 x 16) + 0 = 496, enquanto a de saída tem 17 pois (16 x 1) + 1 = 17.

In [8]:
neural_network.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 16)             │           496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 513 (2.00 KB)

 Trainable params: 513 (2.00 KB)

 Non-trainable params: 0 (0.00 B)

**5) Adicione um otimizador Adam (https://www.tensorflow.org/api_docs/python/tf/keras/optimizers) e especifique a classe loss binário - binary crossentropy (https://www.tensorflow.org/api_docs/python/tf/keras/losses) e a classe metrics para utilizar a métrica de avaliação de acurácia binária (https://www.tensorflow.org/api_docs/python/tf/keras/metrics)**

In [10]:
# Compila rede com função de perda binária e acurácia binária como métrica de validação
neural_network.compile(optimizer='adam', loss='binary_crossentropy', metrics=['binary_accuracy'])

**6) Para que servem os otimizadores? Como o otimizador Adam funciona?**

Os otimizadores têm a função de atualizar os pesos da rede neural ao longo do treinamento, com o objetivo de minimizar a função de perda e melhorar a precisão das previsões.

O algoritmo Adam (Adaptive Moment Estimation) ajusta automaticamente as taxas de aprendizado para cada parâmetro da rede neural combinando duas técnicas clássicas: Momentum e RMSprop. Ele utiliza estimativas da média dos gradientes e da variância dos gradientes para adaptar automaticamente a taxa de aprendizado para cada peso individual do modelo. Essa abordagem proporciona uma convergência mais rápida e estável, tornando o Adam um dos métodos de otimização mais utilizados em problemas de aprendizado profundo.

**7) Depois de estruturados os parâmtros da RNA, utilize a função .fit para fazer o treinamento da rede. Como foi utilizado o otimizador Adam, e ele é baseado na descida do gradiente estocástica, é possível definir a quantidade de registros que serão enviados para a RNA, isto é, em cada batch serão utilizados 10 registros. Quantos batches serão utilizados ao total?**

Como o número de registros do conjunto de treinamento é 426 e o `batch_size` é 10, o total de batches **por época** pode ser dado por 426 / 10 ≈ 43. Finalmente, considerando que o treinamento definiu o valor arbitrário de 100 épocas, serão utilizados 43 × 100 = 4300 batches ao longo de todo o treinamento.

**8) Por fim, defina o número de épocas em que ocorre o treinamento igual a 100.**

In [11]:
# Realiza treinamento do modelo com 100 épocas
neural_network.fit(X_train, y_train, batch_size=10, epochs=100)

Epoch 1/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 2s 21ms/step - binary_accuracy: 0.6502 - loss: 132.8360
Epoch 2/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6221 - loss: 74.9856
Epoch 3/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6009 - loss: 61.2179
Epoch 4/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6385 - loss: 38.0046
Epoch 5/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6033 - loss: 26.5005
Epoch 6/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5915 - loss: 22.8324
Epoch 7/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5915 - loss: 19.2905
Epoch 8/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6268 - loss: 14.4833
Epoch 9/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6244 - loss: 13.0635
Epoch 10/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6573 - loss: 10.9621
Epoch 11/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accura

**9) Crie uma variável chamada previsoes para realizar a previsão dos dados de teste (X_teste) O resultado da rede deve ser um valor entre 0 e 1. Porque isso acontece?** <br>

Como estamos usando a função de ativação sigmoid na camada de saída da rede, o resultado da previsão será uma probabilidade entre 0 (0%) e 1 (100%). Ainda que o problema seja de classificação binária, o threshold de 0.5 só é utilizado após previsão, de modo que saídas > 0.5 = 1 e saídas <= 0.5 = 0. Isto porque, se não fosse um problema do gênero, a utilização de outro threshold permitira dividir em mais classes o resultado.

In [12]:
# Realiza previsão para conjunto de teste
predictions = neural_network.predict(X_test)

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 54ms/step


**10) Converta o resultado que está em probabilidade para valores binários com um limiar (treshold) de 0.5**.

In [15]:
# Conversão das previsões para classes binárias: 0 (False) ou 1 (True)
predictions = predictions > 0.5

In [16]:
predictions

array([[False],
       [ True],
       [ True],
       [ True],
       [False],
       [ True],
       [ True],
       [ True],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [ True],
       [ True],
       [False],
       [ True],
       [ True],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [False],
       [ True],
       [ True],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [ True],
       [False],
       [ True],
       [False],
       [ True],
       [ True],
       [ True],
       [False],
       [ True],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [ True],
       [ True],
       [ True],
       [

**9) Explique o resultado da RNA.** <br>

A RNA, apesar de ter utilizado 100 épocas para treinamento, apresentou um resultado intermediário de acurácia, de 86,71%, e perda, de 2. Provavelmente isso se deve à arquitetura da rede, que utiliza apenas uma camada oculta (Perceptron). A taxa relativamente baixa de perda revela que o otimizador foi importante durante o processo de treinamento, ainda que talvez não tenha a configuração ideal para alcançar melhor eficiência da rede.

In [17]:
neural_network.evaluate(X_test, y_test)

5/5 ━━━━━━━━━━━━━━━━━━━━ 1s 71ms/step - binary_accuracy: 0.8671 - loss: 2.0709


[2.070936918258667, 0.867132842540741]

## Camadas e Otimização da RNA

**10) Adicione mais uma camada para oculta densa com 16 neurônio com a função de ativação relu e inicialize os pesos de utilize o Random uniform initializer. Qual o total de parâmetros da RNA agora?**<br>

O total de parâmetros agora é de 785.

In [18]:
# Recriação da rede com uma camada oculta a mais
neural_network = Sequential([
    Input(shape=(ENTRY_LAYER_SIZE,)),

    # Duas camadas ocultas com 16 neurônios
    Dense(units=HIDDEN_LAYER_SIZE, activation='relu', kernel_initializer=initializer),
    Dense(units=HIDDEN_LAYER_SIZE, activation='relu', kernel_initializer=initializer),

    Dense(units=1, activation='sigmoid')
])

classifier = KerasClassifier(
    model=neural_network,
    verbose=0
)

neural_network.summary()


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_2 (Dense)                 │ (None, 16)             │           496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 16)             │           272 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 785 (3.07 KB)

 Trainable params: 785 (3.07 KB)

 Non-trainable params: 0 (0.00 B)

**11) Abaixo são adicionados os parâmetros do otimizador, que são a taxa de aprendizado e o clipvalue. O que eles fazem?** <br>

A taxa de aprendizado (learning rate) determina o tamanho dos ajustes realizados nos pesos da rede a cada atualização. Valores maiores permitem um aprendizado mais rápido, mas podem fazer com que o modelo não converja para a solução ideal. Já valores muito pequenos tornam o treinamento mais lento. No código foi utilizada uma taxa de aprendizado de 0.001, que é um valor comum por proporcionar estabilidade ao treinamento.

O clipvalue é um mecanismo de gradient clipping que limita o valor máximo dos gradientes durante a atualização dos pesos. Com clipvalue = 0.5, qualquer gradiente com magnitude superior a 0.5 é truncado para esse limite. Isso evita o problema de explosão dos gradientes, tornando o treinamento mais estável e reduzindo oscilações excessivas na otimização.

In [19]:
# Compilação da rede com learning rate e clipvalue
optimizer = tf.keras.optimizers.Adam(learning_rate = 0.001, clipvalue = 0.5)
neural_network.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['binary_accuracy'])

**12) Teste novamente a RNA. Aumentar a quantidade de camadas melhorou ou piorou os resultados? Explique o que aconteceu com a RNA e porque.**<br>
Ao aumentar a quantidade de camadas ocultas, a acurácia caiu para 77%, enquanto a taxa de perda (loss) aumentou para aproximadamente 9. Ou seja, a rede apresentou desempenho inferior à anterior. Ao aumentar sua capacidade de aprendizado com mais uma camada de neurônios, a rede se tornou mais suscetível ao **overfitting**, porque criou uma dependência entre os neurônios da primeira camada oculta com os da segunda, de modo que acabasse memorizando ruídos em vez de aprender padrões generalizáveis.

In [20]:
neural_network.fit(X_train, y_train, batch_size=10, epochs=100)

Epoch 1/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 3s 23ms/step - binary_accuracy: 0.5962 - loss: 3869.0811
Epoch 2/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6385 - loss: 197.6100
Epoch 3/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6338 - loss: 121.7810
Epoch 4/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.5869 - loss: 124.5224
Epoch 5/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.6127 - loss: 63.7264
Epoch 6/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - binary_accuracy: 0.6103 - loss: 43.9258
Epoch 7/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.6033 - loss: 57.1935
Epoch 8/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.6385 - loss: 37.9453
Epoch 9/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.6244 - loss: 29.5962
Epoch 10/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.6408 - loss: 34.0371
Epoch 11/100
43/43 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_ac

In [21]:
predictions = neural_network.predict(X_test)

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 52ms/step


In [22]:
neural_network.evaluate(X_test, y_test)

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 50ms/step - binary_accuracy: 0.7762 - loss: 8.9915


[8.991532325744629, 0.7762237787246704]

## K-Fold Cross Validation

In [25]:
# Importa recursos necessários das bibliotecas
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import cross_val_score
from tensorflow.keras.models import Sequential
from tensorflow.keras import backend as k
from tensorflow.keras.layers import Dense, Dropout, InputLayer

In [26]:
'''
Função que cria e compila uma rede neural com as mesmas configurações da anterior:
  - 1 camada de entrada com 30 neurônios
  - 2 camadas ocultas densas com 16 neurônios e função ReLU
  - 1 camada de saída com função sigmoid

Foi adicionado um parâmetro opcional de dropout para reutilização em outras
questões.
'''
def create_network(dropout=False):
    k.clear_session()

    neural_network = Sequential()

    # Camada de entrada
    neural_network.add(InputLayer(shape=(30,)))

    # 1ª camada oculta
    neural_network.add(
        Dense(
            units=16,
            activation='relu',
            kernel_initializer='random_uniform'
        )
    )

    # Dropout entre camadas ocultas
    if dropout:
        neural_network.add(Dropout(0.2))

    # 2ª camada oculta
    neural_network.add(
        Dense(
            units=16,
            activation='relu',
            kernel_initializer='random_uniform'
        )
    )

    # Dropout antes da camada de saída
    if dropout:
        neural_network.add(Dropout(0.2))

    # Camada de saída
    neural_network.add(
        Dense(
            units=1,
            activation='sigmoid'
        )
    )

    optimizer = tf.keras.optimizers.Adam(
        learning_rate=0.001,
        clipvalue=0.5
    )

    neural_network.compile(
        optimizer=optimizer,
        loss='binary_crossentropy',
        metrics=['binary_accuracy']
    )

    return neural_network

**13) Assista o vídeo https://youtu.be/RczbeFs_WbQ?si=f1Yu4LZaGDfZZpbZ para compreender melhor sobre como funciona o k-fold cross validation. Explique como esta rede foi configurada e como é possível chegar no resultado dela. O que é necessário fazer?**<br>
A rede é criada seguindo o mesmo padrão anterior, ou seja:
1. Uma camada de entrada com 30 neurônios
2. Duas camadas ocultas com 16 neurônios
3. Uma camada de saída que utiliza sigmoide como função de ativação

Além de utilizar o otimizador Adam, `batch_size` com valor 10 e 100 épocas de treinamento.

Em relação ao K-Fold, é necessário chamar a função **cross_val_score** que recebe como paramêtro `cv`, responsável por definir o total de folds. Na implementação atual é definido cv=10. Portanto, em cada execução, 9 partes são utilizadas para treinamento e 1 parte é utilizada para validação. Esse processo é repetido 10 vezes, alternando a parte utilizada para teste, de modo que todas as amostras sejam avaliadas uma vez.



In [51]:
# Cria classificador com nova função de criação da rede, sem dropout
neural_network = KerasClassifier(model = create_network(dropout=False), epochs = 100, batch_size = 10)

In [52]:
# Calcula cross validation com 10 folds
resultados = cross_val_score(estimator = neural_network, X = X, y = y, cv = 10, scoring = 'accuracy')

Epoch 1/100


/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 14ms/step - binary_accuracy: 0.5996 - loss: 1.2818
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7090 - loss: 0.5452
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8008 - loss: 0.4572
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8184 - loss: 0.4336
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8262 - loss: 0.4470
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8457 - loss: 0.3803
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8164 - loss: 0.4577
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8398 - loss: 0.4018
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8242 - loss: 0.4936
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8340 - loss: 0.4815
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8203 - loss: 0.53

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 13ms/step - binary_accuracy: 0.6074 - loss: 1.1469
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7168 - loss: 0.5567
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7559 - loss: 0.5189
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7539 - loss: 0.6124
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8027 - loss: 0.5974
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8125 - loss: 0.4434
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7852 - loss: 0.4465
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8125 - loss: 0.4559
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8262 - loss: 0.4666
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8516 - loss: 0.3986
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8672 - loss: 0.43

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 3s 18ms/step - binary_accuracy: 0.6191 - loss: 1.2967
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7070 - loss: 0.6021
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7441 - loss: 0.5277
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8047 - loss: 0.4661
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8184 - loss: 0.5141
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8359 - loss: 0.4634
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8184 - loss: 0.5173
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8203 - loss: 0.4594
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7930 - loss: 0.5443
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8242 - loss: 0.4407
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8125 - loss: 0.54

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 13ms/step - binary_accuracy: 0.6309 - loss: 1.0895
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7676 - loss: 0.5167
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7832 - loss: 0.4553
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8516 - loss: 0.3749
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7715 - loss: 0.5728
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8379 - loss: 0.4347
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8340 - loss: 0.3797
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8359 - loss: 0.3870
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - binary_accuracy: 0.8750 - loss: 0.3768
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8066 - loss: 0.5400
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8184 - loss: 0.54

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 12ms/step - binary_accuracy: 0.5879 - loss: 1.3962
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6855 - loss: 0.5626
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7754 - loss: 0.4819
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8223 - loss: 0.4446
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7852 - loss: 0.5044
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8105 - loss: 0.4788
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8242 - loss: 0.4523
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8438 - loss: 0.4054
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8086 - loss: 0.4928
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8242 - loss: 0.4591
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8125 - loss: 0.65

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 12ms/step - binary_accuracy: 0.5879 - loss: 1.2945
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6406 - loss: 0.6442
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7578 - loss: 0.4869
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7754 - loss: 0.4613
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8359 - loss: 0.4055
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8105 - loss: 0.3976
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7910 - loss: 0.4643
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8574 - loss: 0.3684
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8438 - loss: 0.3820
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8691 - loss: 0.3348
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8535 - loss: 0.36

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 12ms/step - binary_accuracy: 0.6035 - loss: 1.2245
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6973 - loss: 0.5489
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7188 - loss: 0.5016
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7539 - loss: 0.5038
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7578 - loss: 0.4717
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8125 - loss: 0.4772
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8301 - loss: 0.3916
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8398 - loss: 0.3989
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8516 - loss: 0.3591
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8555 - loss: 0.4058
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8770 - loss: 0.34

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 13ms/step - binary_accuracy: 0.6387 - loss: 1.0296
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6973 - loss: 0.5835
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - binary_accuracy: 0.7461 - loss: 0.4751
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.7949 - loss: 0.4503
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.7949 - loss: 0.4470
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8086 - loss: 0.4532
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8027 - loss: 0.5143
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - binary_accuracy: 0.8340 - loss: 0.4053
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8418 - loss: 0.4044
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8047 - loss: 0.4519
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8105 - loss: 0.43

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 13ms/step - binary_accuracy: 0.5801 - loss: 1.2376
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7051 - loss: 0.5497
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8027 - loss: 0.4430
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8125 - loss: 0.4172
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8691 - loss: 0.3774
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8223 - loss: 0.4420
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8320 - loss: 0.4527
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8398 - loss: 0.4070
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8652 - loss: 0.3867
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8438 - loss: 0.4204
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8145 - loss: 0.52

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 13ms/step - binary_accuracy: 0.6277 - loss: 1.2300
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.7583 - loss: 0.5131
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8031 - loss: 0.4438
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7778 - loss: 0.4841
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8324 - loss: 0.4057
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8285 - loss: 0.4183
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8577 - loss: 0.3852
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8596 - loss: 0.4309
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8324 - loss: 0.4759
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8616 - loss: 0.3993
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8538 - loss: 0.38

In [53]:
# Exibe resultados da cross validation
resultados

array([0.73684211, 0.75438596, 0.87719298, 0.71929825, 0.84210526,
       0.85964912, 0.75438596, 0.8245614 , 0.78947368, 0.875     ])

**14) Calcule também o Desvio Padrão dos resultados para avaliar o modelo. O que é possível concluir com esse resultado?**

O desvio padrão calculado foi de 7,35%, o que significa que o modelo ainda apresenta grande variação de eficiência a depender do conjunto utilizado para validação. Isso pode ser observado pela divergência entre a acurácia do penúltimo fold (70,1%) com a do anterior a ele (92,9%), o que aponta para um overfitting do modelo e baixa capacidade de generalização.

In [54]:
import numpy as np

# Calcula desvio padrão do K-Fold
standard_deviation = np.std(resultados)

# Exibe desvio padrão
standard_deviation

np.float64(0.05673046826483588)

## Overfitting e Dropout

**14) Como vimos, o modelo anterior sofreu overfitting. Para reduzir esse problema, aplicamos a técnica de regularização conhecida como dropout (artigo oficial: https://www.cs.toronto.edu/~rsalakhu/papers/srivastava14a.pdf e explicação em português: https://www.deeplearningbook.com.br/como-funciona-o-dropout/ ). Aplique o dropout de 20% na primeira e segunda camada oculta. O que acontece com os resultados? E o Desvio Padrão?**

In [55]:
# Recria o modelo dessa vez com camadas de Dropout de 20%
model = create_network(dropout=True)
neural_network = KerasClassifier(model=model, epochs=100, batch_size=10)

In [56]:
# Recalcula cross validation
results = cross_val_score(estimator=neural_network, X=X, y=y, cv=10, scoring='accuracy')

Epoch 1/100


/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 4s 27ms/step - binary_accuracy: 0.5312 - loss: 2.4938
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6289 - loss: 0.6428
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6719 - loss: 0.5812
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6504 - loss: 0.6361
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6855 - loss: 0.5549
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6875 - loss: 0.6252
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7461 - loss: 0.5965
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7266 - loss: 0.5899
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7188 - loss: 0.5686
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7402 - loss: 0.5492
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7695 - loss: 0.53

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 3s 26ms/step - binary_accuracy: 0.5234 - loss: 2.1398
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5938 - loss: 0.6691
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6270 - loss: 0.6197
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6504 - loss: 0.5751
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7070 - loss: 0.5668
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7578 - loss: 0.5401
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7207 - loss: 0.5862
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7441 - loss: 0.5827
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7480 - loss: 0.5118
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7715 - loss: 0.5113
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7754 - loss: 0.51

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 3s 25ms/step - binary_accuracy: 0.5430 - loss: 2.1173
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6309 - loss: 0.7132
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6582 - loss: 0.6191
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6602 - loss: 0.6484
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7051 - loss: 0.5756
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7422 - loss: 0.5935
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7539 - loss: 0.5030
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7656 - loss: 0.5471
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7734 - loss: 0.6196
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7402 - loss: 0.5994
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7617 - loss: 0.58

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 3s 25ms/step - binary_accuracy: 0.5410 - loss: 2.2131
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6465 - loss: 0.6627
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6621 - loss: 0.6763
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7305 - loss: 0.5379
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7227 - loss: 0.5654
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7402 - loss: 0.5560
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7754 - loss: 0.5195
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7598 - loss: 0.5835
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7695 - loss: 0.6302
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7754 - loss: 0.5428
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8105 - loss: 0.54

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 3s 26ms/step - binary_accuracy: 0.5371 - loss: 1.9828
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6172 - loss: 0.6241
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6133 - loss: 0.6052
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5820 - loss: 0.6389
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6426 - loss: 0.5768
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6914 - loss: 0.5750
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7070 - loss: 0.5994
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7188 - loss: 0.6109
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7188 - loss: 0.6163
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7051 - loss: 0.6469
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7383 - loss: 0.54

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 4s 29ms/step - binary_accuracy: 0.5469 - loss: 2.5122
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - binary_accuracy: 0.6426 - loss: 0.6343
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6484 - loss: 0.6699
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7520 - loss: 0.5031
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7578 - loss: 0.5609
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7969 - loss: 0.4547
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7891 - loss: 0.4827
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7598 - loss: 0.5595
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8086 - loss: 0.4953
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8008 - loss: 0.4916
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8027 - loss: 0.49

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 3s 26ms/step - binary_accuracy: 0.5859 - loss: 1.7350
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5957 - loss: 0.6983
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6797 - loss: 0.5737
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7188 - loss: 0.5359
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7168 - loss: 0.5789
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7227 - loss: 0.5818
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7832 - loss: 0.5124
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7754 - loss: 0.4895
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7539 - loss: 0.5633
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7578 - loss: 0.5861
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7637 - loss: 0.55

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 3s 26ms/step - binary_accuracy: 0.5430 - loss: 1.8956
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6055 - loss: 0.6223
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6406 - loss: 0.6235
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6836 - loss: 0.5375
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7324 - loss: 0.5509
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7148 - loss: 0.5641
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7637 - loss: 0.5109
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7324 - loss: 0.5245
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7480 - loss: 0.5532
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7734 - loss: 0.4987
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7930 - loss: 0.55

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 4s 26ms/step - binary_accuracy: 0.5234 - loss: 2.1012
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6016 - loss: 0.6286
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6445 - loss: 0.6194
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6738 - loss: 0.5818
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7324 - loss: 0.5511
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7598 - loss: 0.5139
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7578 - loss: 0.5220
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7656 - loss: 0.4945
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7676 - loss: 0.5161
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8047 - loss: 0.4902
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7930 - loss: 0.45

/usr/local/lib/python3.12/dist-packages/keras/src/saving/saving_lib.py:797: UserWarning: Skipping variable loading for optimizer 'adam', because it has 14 variables whereas the saved optimizer has 2 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


52/52 ━━━━━━━━━━━━━━━━━━━━ 4s 25ms/step - binary_accuracy: 0.5750 - loss: 1.9245
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6218 - loss: 0.7051
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.5887 - loss: 0.7099
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6179 - loss: 0.5898
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6491 - loss: 0.6015
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.6764 - loss: 0.6022
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6901 - loss: 0.5921
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7290 - loss: 0.5797
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7368 - loss: 0.5930
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7544 - loss: 0.5415
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7758 - loss: 0.50

In [57]:
# Exibe resultados da cross validation com Dropout na rede
results

array([0.66666667, 0.87719298, 0.96491228, 0.80701754, 0.9122807 ,
       0.94736842, 0.87719298, 0.96491228, 0.9122807 , 0.91071429])

In [58]:
import numpy as np
standard_deviation = np.std(results)
standard_deviation

np.float64(0.08524970301375259)

## Tuning dos Hiperparâmetros (técnicas de ajuste)

Hyperparameter Tuning (Ajuste de Hiperparâmetros) é o processo geral de encontrar os melhores valores para os hiperparâmetros de um modelo de machine learning.
Exemplos de hiperparâmetros:

- Taxa de aprendizado (learning rate)

- Número de árvores em um RandomForest

- Número de camadas ou neurônios em uma rede neural

- Tipo de kernel em um SVM

O tuning pode ser feito por diferentes técnicas, como:

- Grid Search

- Random Search

- Bayesian Optimization, etc.

**15) Descreva como a RNA foi configurada para fazer o processo de tuning.**
A rede é criada em partes de maneira igual, mesmas camadas e dropouts. Porém agora a função que cria a rede recebe os hiperparamêtros. Criamos **params** que serão testados utilizando Grid Search, optamos por utilizar a segunda versão do params que conta com menos opções pois a primeira acaba crashando a execução após algumas épocas de treinamento.<br>

Os melhores paramêtros encontrados foram: <br>
1. batch_size: 10
2. epochs: 50
3. model__activation: 'relu'
4. model__kernel_initializer: random_uniform
5. model__loss: binary_crossentropy
6. model__neurons: 16
7. model__optimizer: 'adam'

E que com eles foi possível obter uma acurácia de aproximadamente 89%.


In [ ]:
def create_network(optimizer, loss, kernel_initializer, activation, neurons):
  k.clear_session()
  rede_neural = Sequential([
      tf.keras.layers.InputLayer(shape=(30,)),
      tf.keras.layers.Dense(units=neurons, activation=activation, kernel_initializer=kernel_initializer),
      tf.keras.layers.Dropout(rate = 0.2),
      tf.keras.layers.Dense(units=neurons, activation=activation, kernel_initializer=kernel_initializer),
      tf.keras.layers.Dropout(rate = 0.2),
      tf.keras.layers.Dense(units=1, activation = 'sigmoid')])
  rede_neural.compile(optimizer = optimizer, loss = loss, metrics = ['binary_accuracy'])
  return rede_neural

In [ ]:
neural_network = KerasClassifier(model=create_network)

In [ ]:
params = {
    'batch_size': [10, 30],
    'epochs': [50, 100],
    'model__optimizer': ['adam', 'sgd'],
    'model__loss': ['binary_crossentropy', 'hinge'],
    'model__kernel_initializer': ['random_uniform', 'normal'],
    'model__activation': ['relu', 'tanh'],
    'model__neurons': [16, 8]
}

In [ ]:
#caso tiver problema com a configuração anterior, testar essa
params = {
    'batch_size': [10, 30],
    'epochs': [50],
    'model__optimizer': ['adam'],
    'model__loss': ['binary_crossentropy'],
    'model__kernel_initializer': ['random_uniform', 'normal'],
    'model__activation': ['relu'],
    'model__neurons': [16]
}

In [ ]:
params

{'batch_size': [10, 30],
 'epochs': [50],
 'model__optimizer': ['adam'],
 'model__loss': ['binary_crossentropy'],
 'model__kernel_initializer': ['random_uniform', 'normal'],
 'model__activation': ['relu'],
 'model__neurons': [16]}

In [ ]:
from sklearn.model_selection import GridSearchCV
grid_search = GridSearchCV(estimator = neural_network, param_grid = params,
                          scoring = 'accuracy', cv = 5)

In [ ]:
grid_search = grid_search.fit(X, y)

Epoch 1/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 4s 42ms/step - binary_accuracy: 0.5824 - loss: 1.0185
Epoch 2/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6308 - loss: 0.6474
Epoch 3/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7033 - loss: 0.5784
Epoch 4/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6703 - loss: 0.5682
Epoch 5/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7341 - loss: 0.4839
Epoch 6/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7407 - loss: 0.5177
Epoch 7/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7670 - loss: 0.4562
Epoch 8/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7780 - loss: 0.4682
Epoch 9/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7824 - loss: 0.4449
Epoch 10/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8286 - loss: 0.4106
Epoch 11/50
46/46 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8132 - loss: 0.3

In [ ]:
print(grid_search)

GridSearchCV(cv=5,
             estimator=KerasClassifier(model=<function create_network at 0x7d1fed61aac0>),
             param_grid={'batch_size': [10, 30], 'epochs': [50],
                         'model__activation': ['relu'],
                         'model__kernel_initializer': ['random_uniform',
                                                       'normal'],
                         'model__loss': ['binary_crossentropy'],
                         'model__neurons': [16], 'model__optimizer': ['adam']},
             scoring='accuracy')


In [ ]:
best_params = grid_search.best_params_
print(best_params)

{'batch_size': 10, 'epochs': 50, 'model__activation': 'relu', 'model__kernel_initializer': 'random_uniform', 'model__loss': 'binary_crossentropy', 'model__neurons': 16, 'model__optimizer': 'adam'}


In [ ]:
best_score = grid_search.best_score_
print(best_score)

0.8910883403198261


**16) É possível melhorar ainda mais a curácia da RNA? Como?**

## Para salvar o seu modelo de RNA

In [ ]:
best_model = grid_search.best_estimator_
best_model.model_.save('classificador_breast.keras')


In [ ]:
classificador_novo = tf.keras.models.load_model('/content/classificador_breast.keras')

In [ ]:
classificador_novo.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 16)             │           496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 16)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 16)             │           272 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 16)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,357 (9.21 KB)

 Trainable params: 785 (3.07 KB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 1,572 (6.14 KB)